testnet/stacks-node/src/nakamoto_node/signer_coordinator.rs

```diff
@@ -336,10 +377,49 @@ impl SignerCoordinator {
                         return Err(NakamotoNodeError::BurnchainTipChanged);
                     }
 
+                    if rejections_timer.elapsed() > *rejections_timeout {
+                        warn!("Timed out while waiting for responses from signers";
+                                  "elapsed" => rejections_timer.elapsed().as_secs(),
+                                  "rejections_timeout" => rejections_timeout.as_secs(),
+                                  "rejections" => rejections,
+                                  "rejections_threshold" => self.total_weight.saturating_sub(self.weight_threshold)
+                        );
+                        return Err(NakamotoNodeError::SigningCoordinatorFailure(
+                            "Gave up while trying reaching the threshold".into(),
+                        ));
+                    }
+
                     continue;
                 }
             };
 
+            if rejections != block_status.total_reject_weight as u64 {
+                rejections = block_status.total_reject_weight as u64;
+                let rejections_timeout_tuple = block_rejection_timeout_steps
+                    .range((Included(0), Included(rejections)))
+                    .last()
+                    .ok_or_else(|| {
+                        NakamotoNodeError::SigningCoordinatorFailure(
+                            "Invalid rejection timeout step function definition".into(),
+                        )
+                    })?;
+                rejections_timeout = rejections_timeout_tuple.1;
+                info!("Number of received rejections updated, resetting timeout";
+                                    "rejections" => rejections,
+                                    "rejections_timeout" => rejections_timeout.as_secs(),
+                                    "rejections_threshold" => self.total_weight.saturating_sub(self.weight_threshold));
+                rejections_timer = Instant::now();
```

Comment:
I think its better not to update the `rejections_timer`. That way the heuristic applies to the total time that the miner is waiting for responses. This means that sometimes a rejection's arrival would immediately cause the timeout to be reached.

Reply:
Yeah, it makes sense, and should be safe given that in the case of a 'quick timeout' the loop will start again from scratch by reproposing the block (unless the tip changed) and in such a case the  rejections_timer will be reset